### PR TITLE
[Captcha] Implement failed logins ceiling

### PR DESCRIPTION
**Pull request by vincentsalucci**
## Type of change
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Code changes
- **GlobalSettings.cs**: Added `MaximumFailedLoginAttempts` into `Captcha` object which acts as the ceiling for triggering captcha validation
- **Entities/User.cs**: Added `FailedLoginCount` and `LastFailedLoginDate`
- **BaseRequestValidator**: Added `UserRepository` and `CaptchaValidationService` into constructor. Added `ResetFailedAuthDetailsAsync` and `UpdateFailedAuthDetailsAsync` that will handle resetting, incrementing, and emailing the user. Passed along `unknownDevice` state in order to prevent multiple db hits unnecessarily.
- **CustomTokenRequestValidator**: Updated constructor
- **ResourceOwnerRequestValidator**: Updated constructor. Passed `user.FailedLoginCount` to the captcha required check. Passed in `unknownDevice` status to the parent `ValidateAsync` method
- **Handlebars/FailedLoginAttemps**: Created requested email templates
- **Handlebars/FailedTwoFactorAttemps**: Created requested email templates
- **FailedAuthAttemptsModel**: Created model based on the existing `NewDeviceLoggedInModel`
- **ICaptchaValidationService**: Updated `RequireCaptchaValidation` to take an additional "nullable" parameter. Created `ValidateFailedAuthEmailConditions` to check for sending the necessary emails in regards to captcha protection. 
- **IMailService**: Added `SendFailedLoginAttemptsEmailAsync` and `SendFailedTwoFactorAttemptsEmailAsync` methods
- **HCaptchaValidationService**: Added additional `or` conditional for checking User's `FailedLoginCount` and comparing it to the `MaximumFailedLoginAttempts` constant. Added `ValidateFailedAuthEmailConditions` for determining if the email is necessary to send (Note: this was added to the service to prevent checking additional conditions in the `BaseRequestValidator` class while also NoOp'ing out if Captcha isn't setup)
- **HandlebarsMailService**: Added `SendFailedLoginAttemptsEmailAsync` and `SendFailedTwoFactorAttemptsEmailAsync` methods for creating and sending specific captcha required emails
- **NoopMailService**: Added noop methods
- **NoopCaptchaValidationService**: Added noop methods
- **DB**: Updated tables, stored procedures, views, and migration scripts (EF included)
- **appSettings.Production.json**: Set default for `MaximumFailedLoginAttempts` to 5
- **appSettings.SelfHosted.json**: Set default for `MaximumFailedLoginAttempts` to `null`

## Testing requirements
- After hitting the failed logins ceiling, make sure that captcha validation is required before a successful login.

## Before you submit
- [X] I have checked for formatting errors (`dotnet tool run dotnet-format --check`) (required)
- [X] If making database changes - I have also updated Entity Framework queries and/or migrations (TODO!!!!)
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [X] This change requires a **documentation update** (notify the documentation team)
- [X] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

**Comment by vincentsalucci** [comment]
@bitwarden/dept-devops - We will need an environment variable to signify the default value of `MaximumFailedLoginAttempts` (name pending) for our production cloud environment

**Comment by vincentsalucci** [comment]
@fschillingeriv - Will need to update documentation in regards to the feature implemented in this PR.

**Comment by mimartin12** [comment]
@vincentsalucci Would this be something that could be implemented inside a config file? Such as https://github.com/bitwarden/server/blob/master/src/Api/appsettings.json?
The environment variable would work, but it would be more manageable inside the existing config files.

**Comment by vincentsalucci** [comment]
> @vincentsalucci Would this be something that could be implemented inside a config file? Such as https://github.com/bitwarden/server/blob/master/src/Api/appsettings.json? The environment variable would work, but it would be more manageable inside the existing config files.

@kspearrin - Thoughts?

File: src/Core/IdentityServer/BaseRequestValidator.cs [hunk 1]
```diff
@@ -502,5 +511,32 @@ private async Task<Device> SaveDeviceAsync(User user, ValidatedTokenRequest requ
...
+        private async Task ResetFailedAuthDetailsAsync(User user)
+        {
+            // Early escape if db hit not necessary
+            if (user.FailedLoginCount == 0)
+            {
+                return;
+            }
+
+            user.FailedLoginCount = 0;
+            user.RevisionDate = user.AccountRevisionDate = DateTime.UtcNow;
```
Comment (kspearrin) [inline src/Core/IdentityServer/BaseRequestValidator.cs:524]:
```suggestion
            user.RevisionDate = DateTime.UtcNow;
```

**Comment by kspearrin** [comment]
> > @vincentsalucci Would this be something that could be implemented inside a config file? Such as https://github.com/bitwarden/server/blob/master/src/Api/appsettings.json? The environment variable would work, but it would be more manageable inside the existing config files.
> 
> @kspearrin - Thoughts?

Yes, we could put some defaults there.

File: src/Core/IdentityServer/BaseRequestValidator.cs (hunk 1 above)
Comment (vincentsalucci) [reply src/Core/IdentityServer/BaseRequestValidator.cs]: Updated!

File: src/Core/IdentityServer/BaseRequestValidator.cs [hunk 2]
```diff
@@ -502,5 +511,32 @@ private async Task<Device> SaveDeviceAsync(User user, ValidatedTokenRequest requ
...
+            await _userRepository.ReplaceAsync(user);
+        }
+
+        private async Task UpdateFailedAuthDetailsAsync(User user, bool twoFactorInvalid, bool unknownDevice)
+        {
+            var utcNow = DateTime.UtcNow;
+            user.FailedLoginCount = ++user.FailedLoginCount;
+            user.LastFailedLoginDate = utcNow;
```
[the remark this answers is not in the document]
Comment (vincentsalucci) [reply src/Core/IdentityServer/BaseRequestValidator.cs]: Updated!

File: src/Core/IdentityServer/BaseRequestValidator.cs [hunk 3]
```diff
@@ -502,5 +511,32 @@ private async Task<Device> SaveDeviceAsync(User user, ValidatedTokenRequest requ
...
+            user.RevisionDate = user.AccountRevisionDate = utcNow;
```
[the remark this answers is not in the document]
Comment (vincentsalucci) [reply src/Core/IdentityServer/BaseRequestValidator.cs]: Good catch - thank you!

File: src/Core/Services/Implementations/UserService.cs [hunk 4]
```diff
@@ -1400,5 +1400,17 @@ public async Task<bool> VerifySecretAsync(User user, string secret)
                 ? await VerifyOTPAsync(user, secret)
                 : await CheckPasswordAsync(user, secret);
         }
+
+        public async Task SendFailedAuthEmailAsync(string email, bool twoFactorInvalid, DateTime utcNow, string ip)
```
[the remark this answers is not in the document]
Comment (vincentsalucci) [reply src/Core/Services/Implementations/UserService.cs]: My original intent was to not have logic inside the `mail service` for deciding which email to send, but that was before I rewrote these methods and had all necessary data available within the `baseRequestValidtor` class. I've removed the updates to the `User Service` and just added the logic into the existing method. 👍

**Comment by vincentsalucci** [comment]
@mimartin12 After a chat with @kspearrin - we can definitely take the approach you suggested. Thanks for that 👍

**Review by kspearrin — APPROVED** [review]
Seems good to me, though someone else should probably review your changes for EF migrations

**Review by addisonbeck — APPROVED** [review]

**Pull request merged**